### ISSUE TITLE
Mobx not re-rendering in sub-project

### ISSUE THREAD

**Issue by ptallett** [mobxjs/mobx#2270]
Hi!

We are building a React solution that will be split into several sub-projects to facilitate independent development. We have taken a simple approach to this, we have a shell project and several sub-projects. All projects are created with CRA --typescript. The sub-projects all publish to a private NPM feed and the shell project simply references them from the feed. All the sub-projects can be run standalone.

My issue is we are using mobx and mobx-react in all projects, which works fine when we run the projects standalone, but the components do not re-render when updating observables in the sub-projects when calling them from the shell project. I am using tsc from the command line to package the projects, and can see the decoration being applied, but adding console trace statements show the render methods are not called when we change observables.

I've probably messed something up with dependencies and duplicate copies of React, but I've made sure that everything is using the same versions.

If someone could give me pointers where to hunt, that would be great.

Cheers,
Paul

**Comment by FredyC** [comment]
If you are using Yarn then `yarn why mobx` is a very useful command to discover duplicates. Otherwise, without some sort of reproduction, it's hard to tell what's wrong in your case.

**Comment by mweststrate** [comment]
some versions doesn't dedupe, to efficiently dedupe you'd have to set up
additional config in your projects. Another solution is to make react and
mobx peer deps on all subprojects, and dev dependencies in case they run
stand alone.

**Comment by ptallett** [comment]
Thanks for the help. I've repro'ed in a small project.

https://1drv.ms/u/s!AlBap96oLtEvi4wVG3--bY-9JEaH_w?e=FpFqaK

You will need to create a private feed (mine's in Azure DevOps) and update the two .npmrc files.

Then:
- cd reprosub
- npm install
- npm start
- click the number - it should increment
- npm run package
- npm publish
- cd ../repro
- npm install
- npm run start
- click the number - it doesn't increment.

Cheers,
Paul

**Comment by FredyC** [comment]
Can you please tweak it without the need for Azure? Not only for our convenience, but also to prove the problem is not in that somehow.

**Comment by ptallett** [comment]
It doesn’t need Azure, it just needs an NPM feed. You can create where you like.

**Comment by FredyC** [comment]
I don't follow what is NPM feed, never heard of it.

**Comment by ptallett** [comment]
NPM registry?

**Comment by FredyC** [comment]
That sounds overly complicated. I just did `file:../reprosub` in `package.json`, install deps and it's obvious that both folders have its own `mobx` and `react`. I am pretty sure you need to do what Michel suggested with peer/dev deps to avoid duplicates.

This is actually a good first case of using devDependencies in React app. Never thought about that. Yea, micro-frontends is definitely an interesting approach :)

OT: Oh boy, I thought they have improved NPM speed, but using it after a while it's still like a 🐌 compared to Yarn 😮

**Comment by ptallett** [comment]
Um, I thought I'd done what Michael suggested? I have mobx and react as devDependencies & peerDependencises in reprosub and installed them as deps in repro

**Comment by mweststrate** [comment]
These two pointers pretty much sum up what there is to know about this
subject. https://github.com/mobxjs/mobx/issues/1082#issuecomment-381881926.
Please check those out first. The clue is to make sure you never have
packages twice. There are many ways to achieve that, using peerDeps,
linking to shared dir, pulling node_modules up to a shared direction,
module resolutions in your build config, yarn workspaces, relative-deps,
etc. How you exactly achieve that depends on your setup, so best check out
the above pointers first and verify you indeed deduped (for example if you
use devDeps, but share the bundled output of your subprojects, you still
have two copies).

**Comment by FredyC** [comment]
Well, so I've used my favorite Yalc tool to simulate package installation closer to the real one. Deps are there only once in that case. And yet the result is still not working.

To be honest I am not much of a fan of decorators and classes. So out of curiosity, I tried to rewrite to a functional component and suddently it works 😮 

```
export default observer(function ReproApp() {
    const store = useLocalStore(() => ({
        count: 0
    }))
    console.log('render')
    return (
      <div style={{fontSize: 100}} onClick={() => store.count++}>
        {store.count}
      </div>
    );
  }
)
```

I have no idea why it would be a problem. Someone who knows about decorators need to dig through that.

**Comment by mweststrate** [comment]
sounds more like the babel / ts configs are not consistent across the
projects then? (IIRC mobx will actually show a warning in the dev tools if
you have multiple instances)

**Comment by ptallett** [comment]
Thanks. If you check the dist folder of reprosub, you will see that the decorators do appear to be applied, so I don't think that's the issue.

You notice I am not creating a mobx store - could the two projects be using different stores?

Cheers,
Paul

**Comment by ptallett** [comment]
The tsconfig is definitely not consistent across the projects. We are specifying the options on the package command line and also in a tsconfig file (two different places). What option would cause this type of issue?

**Comment by FredyC** [comment]
It's really strange, looking at the transpiled output of the class, I find somewhat odd.

```js
var ReproApp = /** @class */ (function (_super) {
    __extends(ReproApp, _super);
    function ReproApp() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.count = 0; // <-- instance property
        return _this;
    }
    ReproApp.prototype.render = function () {
        var _this = this;
        console.log('render');
        return (React.createElement("div", { style: { fontSize: 100 }, onClick: function () { return _this.count++; } }, this.count));
    };
    __decorate([
        mobx_1.observable
    ], ReproApp.prototype, "count"); <-- attaching observable on prototype?
    return ReproApp;
}(React.Component));
```

If the `reprosub` wouldn't be working on its own, I would almost guessed this is a culprit, but something is odd when this code accessed from the other app 🤷‍♂

**Comment by ptallett** [comment]
That transpiled output is not used when the project runs standalone AFAIK.

**Comment by mweststrate** [comment]
That is correct, the prop gets inherited from the proto, and then the read
made by the constructor in intercepted.

**Comment by mweststrate** [comment]
ahh.. yeah if there are multiple different build setups, make sure in all
of them the decorators are setup according to the official docs on
mobx.js.org....

**Comment by FredyC** [comment]
Right, when running `subrepo` as an app, it's transpiled by Babel Typescript compared to when running `tsc` over it for packaging. That's most likely the culprit and you need to use the same transpilation method. I am not sure which one is better.

Well, another more reason why not to bother with decorators for me. Full of issues overweights benefits by a long stretch.

**Comment by ptallett** [comment]
ISSUE SOLVED!

I created a separate tsconfig file ->tspackage.json and made sure all the TS options were the same, added input and outdir and then it all worked!

Thanks guys for the super quick responses and great support! We have a demo next week to our stakeholders so brilliant that we can demo this.

I *LOVE* mobx... Kill that Red*x...

Cheers,
Paul

**Comment by FredyC** [comment]
That's really interesting, would be good to know which compiler options have caused it so we can warn about that in docs. I tried to copy tsconfig from `reprosub` to `repro` but it did not help. The only difference was missing `experimentalDecorators` in `repro`. It's probably something else then.

It would be great if you can elaborate more on the actual solution. We can wait after your demo for sure :)

**Comment by ptallett** [comment]
No, that was the point, the packaging step wasn't using tsconfig.json, it only used the command line options that we had specified. The solution was to create a new tspackage.json that extends tsconfig.json, and then call tsc with the -project option rather than other command switches. That way tsc uses all the same settings during packaging as when you run it standalone. It wasn't a mismatch between repro and reprosub, it was that the packaged version of reprosub was broken - it wasn't compiled with the correct TSC options causing the decorations to not work (by the way I *love* decorators - nice and easy to learn and understand. Also I hate React functional components - hard to learn/remember the syntax and understand. :-))

I've updated the repro with the corrected solution in case others come here in the future - it looks like I can drop it on the answer.

A few keywords for the search engines: micro frontends micro front ends micro services.

Cheers,
Paul
[repro.zip](https://github.com/mobxjs/mobx/files/4129402/repro.zip)
